ouroboros-consensus-diffusion/test/consensus-test/Test/Consensus/PeerSimulator/BlockFetch.hs

```diff
@@ -124,32 +126,54 @@ runBlockFetchClient peerId tracer wait fetchClientRegistry controlMsgSTM getCurr
     ntnVersion :: NodeToNodeVersion
     ntnVersion = maxBound
 
+-- TODO Move this to Handlers, create ScheduledBlockFetchServer
 mockBlockFetchServer ::
      forall m.
      (IOLike m)
   => STM m ()
   -> Tracer m String
-  -> m (AnchoredFragment TestBlock)
+  -> m (AnchoredFragment TestBlock, TestBlock)
   -> BlockFetchServer TestBlock (Point TestBlock) m ()
 mockBlockFetchServer wait tracer getCurrentChain = idle
   where
     idle :: BlockFetchServer TestBlock (Point TestBlock) m ()
     idle = BlockFetchServer check ()
 
-    -- REVIEW: When the range does not match the current chain, we block until it doesn't.
-    -- This means that the client will never be able to change the range, like when rolling back.
     check :: ChainRange (Point TestBlock) -> m (BlockFetchBlockSender TestBlock (Point TestBlock) m ())
     check (ChainRange from to) = do
-      curChain <- getCurrentChain
-      case AF.sliceRange curChain from to of
-        Nothing    -> do
-          trace ("Waiting for next tick for range: " ++ condense from)
-          atomically wait
-          trace "unblocked"
-          check (ChainRange from to)
-        Just slice -> do
-          trace ("Sending slice " ++ terseFrag slice)
-          pure (SendMsgStartBatch $ sendBlocks (AF.toOldestFirst slice))
+      (hpChain, bp) <- getCurrentChain
+      -- First, check whether the block point is on the same chain, and before or equal to, the header point.
+      case AF.rollback (blockPoint bp) hpChain of
+        Nothing ->
+          -- REVIEW: Should this be fatal?
+          error "Block point isn't on the same chain as header point"
+          -- pure noBlocks
+        Just bpChain ->
+          -- Next, check whether the requested range is contained in the fragment before the block point.
+          -- REVIEW: this is Nothing if only _part_ of the range is on the current chain.
+          -- Is it correct that we don't want to send anything in this case?
+          case AF.sliceRange bpChain from to of
+            Nothing    -> do
+              trace ("Waiting for next tick for range: " ++ tersePoint from ++ " -> " ++ tersePoint to)
+              -- Finally, if we cannot serve blocks, decide whether to send NoBlocks before yielding control
+              -- to the scheduler.
+              -- If the @to@ point is not part of the chain up to HP, we must have switched to a fork, so we
+              -- need to give the client a chance to adapt the range before trying again with the next tick's
+              -- chain.
+              -- Otherwise, we simply have to wait for BP to advance sufficiently, and we block without sending
+              -- a message, to simulate a slow response.
+              case AF.withinFragmentBounds to hpChain of
+                False -> pure noBlocks
```

Comment:
yeah, though this case only happens when BP doesn't allow these blocks, which would mean that the schedule looks like this:
```
HP 5 | BP 3
HP 5' | BP 4
HP 5' | BP 5
```
where we switch to a fork for HP (5') in tick 2, while only allowing block 4 to be sent, and then allowing block 5 in the next tick.
Assuming then that the client requests the range `[4, 5]` in the second tick, we would either:
* stall, sending both blocks in the next tick
* send NoBlocks, also sending both blocks in the next tick, but allowing for the client to update the range because it has now switched to the fork as well.

The original problem was that BlockFetch would keep re-requesting the same range, keeping the tick from concluding. This would be fixed in either of these two cases, since we end the tick even when NoBlocks is sent.

But if we only stall, it would _required_ that the schedule contains `BP 5` instead of switching to the fork, otherwise the server would never escape from that tick.
Do you think the schedule can guarantee that?

Reply:
oh, isn't it possible that the NUT might have gotten the headers from another peer and is requesting the blocks without having seen them advertised?

Reply:
@amesgen?

I don't know really.